<a href="https://colab.research.google.com/github/endzix22/machine-learning-bootcamp/blob/main/01_linear_regression_normal_equationipynb.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
import numpy as np

nana

uwaga te same dane zostaly wziete co mam w zeszycie 

In [5]:
x1=np.array([1,2,3,4,5,6])
y=np.array([3000,3250,3500,3750,4000,4250])
m=len(x1)

print(f'lata pracy: {x1}')
print(f'wynagrodzenie: {y}')
print(f'liczba probek: {m}')

lata pracy: [1 2 3 4 5 6]
wynagrodzenie: [3000 3250 3500 3750 4000 4250]
liczba probek: 6


In [8]:
#zmienimy rozmiar naszych danych abysmy mogli dalej z nimi pracowac
x1=x1.reshape(m,1)
print(x1)
print(x1.shape)

[[1]
 [2]
 [3]
 [4]
 [5]
 [6]]
(6, 1)


In [13]:
#teraz czas wygenerowac ten wektor 1 w x (spojrz zeszyt)
#powszechnie nazywany jest bias
bias=np.ones((m,1))
print(bias)
print(bias.shape)

[[1.]
 [1.]
 [1.]
 [1.]
 [1.]
 [1.]]
(6, 1)


In [15]:
#polaczenie tego x w calosc z bias
x=np.append(bias,x1,axis=1)
print(x)
print(x.shape)

[[1. 1.]
 [1. 2.]
 [1. 3.]
 [1. 4.]
 [1. 5.]
 [1. 6.]]
(6, 2)


Regresja liniowa w $R^2$:
$$Y = w_0 + w_1X_1$$
$$Y = W^TX$$
gdzie: $$W= \left[\begin{matrix}w_0\\w_1\end{matrix}\right]$$ $$ X= \left[\begin{matrix}1\\X_1\end{matrix}\right] $$ stąd $$ W^T= \left[\begin{matrix}w_0&w_1\end{matrix}\right] $$    
$$Y = W^TX = \left[\begin{matrix}w_0&w_1\end{matrix}\right] \cdot  \left[\begin{matrix}1\\X_1\end{matrix}\right] = w_0 + w_1X_1$$

Równanie normalne - równanie pozwalające obliczyć minimum funkcji straty (o ile istnieje)
$$W = (X^TX)^{-1}(X^TY)$$

WEZMIEMY SIE ZA LICZENIE :
$$W = (X^TX)^{-1}(X^TY)$$

In [17]:
#TERAZ POLICZYMY xT*X
np.dot(x.T,x)

array([[ 6., 21.],
       [21., 91.]])

In [21]:
#teraz czas na (xt*x)^(-1)
L = np.linalg.inv(np.dot(x.T, x))
L

array([[ 0.86666667, -0.2       ],
       [-0.2       ,  0.05714286]])

In [22]:
#teraz liczymy xT*Y
P=np.dot(x.T,y)
P

array([21750., 80500.])

In [24]:
#mnozemy nawias razy nawias otrzymujac W 
np.dot(L,P)

array([2750.,  250.])

Końcowa postać modelu$$Y = 2750 + 250X_1$$

Regresja liniowa przy użyciu biblioteki scikit-learn

In [26]:
from sklearn.linear_model import LinearRegression 
regression=LinearRegression()
#importujemy ta klase Linearregression
regression.fit(x1,y)
#dane x i y sa zdefiniowane na samej gorze notatnika
#nie dodajemy juz na szczescie w x jedynek 

print(regression.intercept_)  
#wspolczynnik b 
print(regression.coef_[0])
#wspolczynnik a 

2750.0
250.0
